### PR TITLE
8301869: Regression ~14% in J2dBench-bimg_misc-* in 21-b5 only on linux-aarch64

### PR DESCRIPTION
**Pull request by jayathirthrao**
Under https://bugs.openjdk.org/browse/JDK-8264846 we moved to -O3 level of gcc optimizations from -O1 level for libawt build. This improved our J2DBench performance numbers in some options considerably.

Recent changes done under https://bugs.openjdk.org/browse/JDK-8299337 causes difference in generated code by gcc and this is resulting in performance regression for bimg_misc-* J2DBench options in our performance servers. Under https://bugs.openjdk.org/browse/JDK-8299337 we have just removed unused variables and it is a cleanup task.

We can force gcc to generate position independent code by using -fpic option.Also i have removed -fgcse-after-reload option for gcc, because this is by default covered under -O3 level of optimization introduced under https://bugs.openjdk.org/browse/JDK-8264846.

With this change bimg_misc-* J2DBench option performance regression is resolved and there are no regression in other options of J2DBench or SwingMark and it is verified in our performance servers.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301869](https://bugs.openjdk.org/browse/JDK-8301869): Regression ~14% in J2dBench-bimg_misc-* in 21-b5 only on linux-aarch64

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [e8d25674](https://git.openjdk.org/jdk/pull/12761/files/e8d25674d8ca3824869f3861d3c2789e42e4c18e)

### Contributors
 * Sergey Bylokhov `<serb@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12761/head:pull/12761` \
`$ git checkout pull/12761`

Update a local copy of the PR: \
`$ git checkout pull/12761` \
`$ git pull https://git.openjdk.org/jdk.git pull/12761/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12761`

View PR using the GUI difftool: \
`$ git pr show -t 12761`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12761.diff">https://git.openjdk.org/jdk/pull/12761.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12761#issuecomment-1446201677)

### CODE REVIEW

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk [hunk 1]
```diff
@@ -126,8 +126,10 @@ endif
 # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45977
 ifeq ($(TOOLCHAIN_TYPE), gcc)
   BUILD_LIBAWT_debug_mem.c_CFLAGS := -w
-  # This option improves performance of MaskFill in Java2D by 20% for some gcc
-  LIBAWT_CFLAGS += -fgcse-after-reload
+  # This option forces gcc to generate position independent code, so
+  # that any cleanup of code related to AWT will not change the
+  # generated code and cause performance issues in J2DBench
+  LIBAWT_CFLAGS += -fpic
```
Comment (erikj79) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:132]: We compile with `-fPIC` as a standard flag for position independent code (I just verified that LIBAWT also gets this flag today). My understanding is that `-fpic` may result in smaller/faster code. Overriding with `-fpic` seems like its helping here, but the comment needs to be fixed to explain the actual difference and reason.

Comment (jayathirthrao) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk]: Thanks @erikj79 for your inputs. I have updated the comment to be relevant to the problem it is resolving.

**Review by erikj79 — APPROVED** [review]

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk [hunk 2]
```diff
@@ -126,8 +126,8 @@ endif
 # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45977
 ifeq ($(TOOLCHAIN_TYPE), gcc)
   BUILD_LIBAWT_debug_mem.c_CFLAGS := -w
-  # This option improves performance of MaskFill in Java2D by 20% for some gcc
-  LIBAWT_CFLAGS += -fgcse-after-reload
+  # This option improves performance of FillRect/FillOval in Java2D by 15% for some gcc
+  LIBAWT_CFLAGS += -fpic
```
Comment (prrace) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:130]: So .. acc. to the eval in JBS we can get rid of -fgcse-after-reload because it is implied by -O3
And Erik says -fPIC is implied by -fpic 

So how is -fpic different than -fPIC in a way that helps here ?
So far as I can tell the -fpic are using just limits the size of the global offset table that -fPIC will generate.
Since we had an unlimited offset table before, then -fpic won't change anything. 
So what's the explanation ?

Comment (jayathirthrao) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:130]: Thanks @prrace for your inputs.
I dont think -fPIC is implied by -fpic. @erikj79  please clarify.

Difference between -fpic and -fPIC:
1) With -fPIC we have unlimited storage capacity for global offset table and with -fpic on some platforms we have limitation on this size. For aarch64 it is 28k, but in case of x86 size of GOT is unlimited irrespective of -fPIC or -fpic. This is already captured.
2) Also it looks like if we use -fpic option instead of -fPIC we may generate smaller/faster code as mentioned at http://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html. Looks like in our case, 28k GOT size suffices and we are able to create smaller/faster code for libawt in case of aarch64. This inturn looks to be related to usage of -msmall-data or -mlarge-data flags. -fpic implies -msmall-data and -fPIC implies -mlarge-data while generating code for shared libraries. -msmall-data also talks about data being nearer and easily accessible compared to  -mlarge-data(https://man7.org/linux/man-pages/man1/gcc.1.html)

Also i can make this change specific to aarch64 since we are seeing its effect only on aarch64.

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk [hunk 3]
```diff
@@ -126,8 +126,8 @@ endif
 # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=45977
```
[the remark this answers is not in the document]
Comment (jayathirthrao) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk]: Updated

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk (hunk 2 above)
Comment (erikj79) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:130]: I think you have read more about the differences between -fpic and -fPIC than I have, so nothing I can add here.

Making this option only on aarch64 seems like a reasonable idea if that's where we see the effect.

Comment (prrace) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:130]: > I dont think -fPIC is implied by -fpic. @erikj79 please clarify.

That isn't what I said.
I said they they do very similar things, based on the references I read
with just a specific difference is that lower case "fpic" can be limited in size.

which is why I  wrote
>  So far as I can tell the -fpic are using just limits the size of the global offset table that -fPIC will generate.

My comment
> Since we had an unlimited offset table before, then -fpic won't change anything.

I was trying to point out that if your eval + fix is explained by saying we no longer
have "-fpic" then that's wrong. We never had -fpic.
We didn't change compiler options, but clearly performance changed


The man page you just linked does explain how the mechanism -fpic uses
which limits size could result in better performance

" When generating code for shared libraries, -fpic implies
           -msmall-data and -fPIC implies -mlarge-data."

"When -msmall-data is used,
           objects 8 bytes long or smaller are placed in a small data
           area (the ".sdata" and ".sbss" sections) and are accessed via
           16-bit relocations off of the $gp register."
 
So on some architectures it can be faster.

BUT since we were using -fPIC for a long time, I am not seeing how  you've
explained WHY we now need to explicitly use -fpic to get the performance back ?
In other words "we mysteriously lost performance, but I've found an option that helps".

So possibly, if you applied -fpic to the JDK before the regression it would get even faster ...

**Comment by mrserb** [comment]
When I mention that option in the JBS, I meant that we should take a look on how the methods in the library are "packed" since for some reason "-fpic" make some difference, probably we should use "static" more often or something like that. It is better to check what is the actual difference in the generated/packed code.

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk (hunk 2 above)
Comment (jayathirthrao) [reply make/modules/java.desktop/lib/Awt2dLibraries.gmk:130]: Thanks @prrace. I will analyse further why -fpic is causing change in behaviour in our case.

**Comment by jayathirthrao** [comment]
> When I mention that option in the JBS, I meant that we should take a look on how the methods in the library are "packed" since for some reason "-fpic" make some difference, probably we should use "static" more often or something like that. It is better to check what is the actual difference in the generated/packed code.

Thanks @mrserb for clarification. I will get back with more details.
